File: 8/14.py
def main(n, m):
    result = 0
    numbers: list[int] = list(range(1, n + 1))

    def dfs(level, j, combination):
        nonlocal result
        if level == m:
            print(combination)
            result += 1
            return

        for i in range(j, n):
            combination.append(numbers[i])
            dfs(level + 1, i + 1, combination)
            combination.pop()

    dfs(0, 0, [])

    return result


print(main(4, 2))

File: 8/14.js
// 시간 복잡도: O(nCm)
// 공간 복잡도: O(n)
function main(n, m) {
  let result = 0;
  const numbers = Array.from({ length: n }, (_, i) => i + 1);

  function dfs(level, j, combination) {
    if (level === m) {
      console.log(combination);
      result++;
      return;
    }

    for (let i = j; i < n; i++) {
      combination.push(numbers[i]);
      dfs(level + 1, i + 1, combination);
      combination.pop();
    }
  }

  dfs(0, 0, []);

  return result;
}

console.log(main(4, 2));
